File: research-service/src/angelmind/rehearsal.py
"""Network-free task rehearsal with deterministic estimates only."""

from __future__ import annotations

from dataclasses import dataclass

from .contracts import GovernanceTier, Scope, Task
from .guardrails import GuardrailDecision, evaluate_task


@dataclass(frozen=True, slots=True)
class RehearsalResult:
    tasks: tuple[Task, ...]
    estimated_cost_cents: int
    estimated_duration_minutes: int
    network_calls: int
    tool_executions: int
    decision: GuardrailDecision


def rehearse(scope: Scope, hostname: str, budget_remaining_cents: int, session_remaining_minutes: int) -> RehearsalResult:
    task = Task(
        workspace_id=scope.workspace_id,
        hypothesis_id="hypothetical-policy-context",
        title="Parse program rules and build a hypothetical coverage plan",
        tier=GovernanceTier.TIER_1,
        dry_run=True,
    )
    decision = evaluate_task(task, hostname, scope, budget_remaining_cents, session_remaining_minutes)
    return RehearsalResult(
        tasks=(task,),
        estimated_cost_cents=44,
        estimated_duration_minutes=14,
        network_calls=0,
        tool_executions=0,
        decision=decision,
    )

File: server/control-plane/rehearsal.ts
import type { RehearsalResult, RehearsalTask } from "./contracts";
import { evaluatePolicy } from "./guardrails";

export function buildRehearsal(input: {
  target: string;
  allowlist: string[];
  exclusions: string[];
  safeHarbor: string;
  codeOfConduct: string;
  spentCents: number;
  budgetCents: number;
  sessionLimitMinutes: number;
}): RehearsalResult {
  const policy = evaluatePolicy({
    ...input,
    action: "scope_inventory",
    elapsedMinutes: 0,
    dryRun: true,
  });
  const tasks: RehearsalTask[] = [
    {
      id: "rehearsal-policy-review",
      title: "Parse scope, safe-harbor, and code-of-conduct records",
      action: "policy_review",
      tier: "tier1",
      estimatedMinutes: 6,
      estimatedCostCents: 18,
      simulationOnly: true,
    },
    {
      id: "rehearsal-asset-map",
      title: "Build a hypothetical asset graph from declared workspace records",
      action: "scope_inventory",
      tier: "tier1",
      estimatedMinutes: 9,
      estimatedCostCents: 24,
      simulationOnly: true,
    },
    {
      id: "rehearsal-coverage-plan",
      title: "Plan coverage and governance checkpoints without contacting the target",
      action: "coverage_plan",
      tier: "tier1",
      estimatedMinutes: 11,
      estimatedCostCents: 32,
      simulationOnly: true,
    },
  ];

  return {
    networkCalls: 0,
    toolExecutions: 0,
    taskCount: tasks.length,
    estimatedCostCents: tasks.reduce((sum, task) => sum + task.estimatedCostCents, 0),
    estimatedDurationMinutes: tasks.reduce((sum, task) => sum + task.estimatedMinutes, 0),
    tasks,
    policy,
    checkpoint: { state: "rehearsed", version: 1, createdAt: new Date().toISOString() },
  };
}
